### PR TITLE
Custom domains: track validation process

### PR DESCRIPTION
**Pull request by stsewd**
Started putting this code in -ext, but since it doesn't depend on those models, I'm putting it here. And creating a new app to put all domain related code (can't name it cdn, because we already have that app name, but we can rename it once we have migrated all from that app).

This implements https://github.com/readthedocs/readthedocs-ext/issues/490, only the deactivation part, we don't delete domains.
Basically, the user has 30 days to complete the setup, or we will strop retrying, but they can restart the process by saving the domain if they run out of time.

### How to deploy this change

- Run 0091 before releasing the webs
- Run 0092 after releasing the webs

### CODE REVIEW

File: docs/user/custom-domains.rst [hunk 1]
```diff
@@ -110,6 +110,14 @@ this can cause a delay in validating because there is an exponential back-off in
 
     Loading the domain details in the Read the Docs dashboard and saving the domain again will force a revalidation.
 
+The validation process period has ended
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+After you create your domain, you have 30 days to complete the configuration.
```
Comment (agjohnson) [inline docs/user/custom-domains.rst:116]: Just a little clarification that these are custom domains on RTD:

```suggestion
After you add a new custom domain, you have 30 days to complete the configuration.
```

File: docs/user/custom-domains.rst [hunk 2]
```diff
@@ -110,6 +110,14 @@ this can cause a delay in validating because there is an exponential back-off in
...
+After that period has ended, we will stop trying to validate your domain,
+if you still want to complete the configuration,
```
Comment (agjohnson) [inline docs/user/custom-domains.rst:118]:
```suggestion
After that period has ended, we will stop trying to validate your domain.
If you still want to complete the configuration,
```

File: readthedocs/templates/projects/domain_form.html [hunk 3]
```diff
@@ -36,6 +36,15 @@
       {% endblocktrans %}
     </p>
 
+    {% if not domain.skip_validation and not domain.is_valid and domain.validation_process_expired %}
+      <p class="help_text">
+        {% blocktrans trimmed %}
+        The validation process period has ended.
+        Save the domain to restart the process.
```
Comment (agjohnson) [inline readthedocs/templates/projects/domain_form.html:43]: :+1: for now, but I suppose we could come back to this and alter the button too. Like, if the validation period is over, change the button text to "Retry validation" or something similar

**Review by agjohnson — COMMENTED** [review]
Just poking in really quick on the docs/html. I didn't dive in on the :snake:

File: docs/user/custom-domains.rst [hunk 4]
```diff
@@ -110,6 +110,14 @@ this can cause a delay in validating because there is an exponential back-off in
 
     Loading the domain details in the Read the Docs dashboard and saving the domain again will force a revalidation.
 
+The validation process period has ended
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+After you add a new custom domain, you have 30 days to complete the configuration.
+After that period has ended, we will stop trying to validate your domain.
```
Comment (humitos) [inline docs/user/custom-domains.rst:117]:
```suggestion
Once that period has ended, we will stop trying to validate your domain.
```

File: readthedocs/domains/notifications.py [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+from readthedocs.notifications import Notification
+from readthedocs.notifications.constants import REQUIREMENT
+
+
+class PendingCustomDomainValidation(Notification):
+
+    app_templates = "domains"
+    context_object_name = "domain"
+    name = "pending_domain_configuration"
+    subject = "Pending configuration of custom domain"
```
Comment (humitos) [inline readthedocs/domains/notifications.py:12]: It may be good to include the domain in the subject.

File: readthedocs/domains/templates/domains/notifications/pending_domain_configuration_site.html [hunk 6]
```diff
@@ -0,0 +1,7 @@
+{% url "projects_domains_edit" domain.project.slug domain.pk as domain_url %}
+{% if domain.validation_process_expired %}
+  The validation period of your custom domain <a href="{{ domain_url }}">{{ domain.domain }}</a> has ended.
+  Save the domain to restart the process.
+{% else %}
+  The configuration of your custom domain <a href="{{ domain_url }}">{{ domain.domain }}</a> is pending.
```
Comment (humitos) [inline readthedocs/domains/templates/domains/notifications/pending_domain_configuration_site.html:6]: I think this line should communicate better than an action is required from the user.

**Review by humitos — APPROVED** [review]
Looks good to me!

File: readthedocs/domains/templates/domains/notifications/pending_domain_configuration_site.html (hunk 6 above)
Comment (stsewd) [reply readthedocs/domains/templates/domains/notifications/pending_domain_configuration_site.html:6]: Something like `please complete the configuration` at the end?

File: readthedocs/domains/templates/domains/notifications/pending_domain_configuration_site.html [hunk 7]
```diff
@@ -0,0 +1,7 @@
+{% url "projects_domains_edit" domain.project.slug domain.pk as domain_url %}
+{% if domain.validation_process_expired %}
+  The validation period of your custom domain <a href="{{ domain_url }}">{{ domain.domain }}</a> has ended.
+  Save the domain to restart the process.
```
Comment (humitos) [inline readthedocs/domains/templates/domains/notifications/pending_domain_configuration_site.html:4]: Something more direct here will work better:

```suggestion
  Go to <a>domain configuration</a> and click on "Save" to restart the process.
```

File: readthedocs/domains/templates/domains/notifications/pending_domain_configuration_site.html (hunk 6 above)
Comment (humitos) [reply readthedocs/domains/templates/domains/notifications/pending_domain_configuration_site.html:6]: Yes. It could be something like

```suggestion
  The configuration of your custom domain <a href="{{ domain_url }}">{{ domain.domain }}</a> is pending. Go to <a>domain configuration</a> and follow the instructions to complete it.
```
